lib/Target/Lattigo/LattigoEmitter.cpp

```diff
@@ -0,0 +1,312 @@
+  LattigoEmitter emitter(os, &variableNames);
+  LogicalResult result = emitter.translate(*op);
+  return result;
+}
+
+LogicalResult LattigoEmitter::translate(Operation &op) {
+  LogicalResult status =
+      llvm::TypeSwitch<Operation &, LogicalResult>(op)
+          // Builtin ops
+          .Case<ModuleOp>([&](auto op) { return printOperation(op); })
+          // Func ops
+          .Case<func::FuncOp, func::ReturnOp>(
+              [&](auto op) { return printOperation(op); })
+          .Case<arith::ConstantOp>([&](auto op) { return printOperation(op); })
+          // Lattigo ops
+          .Case<RLWENewEncryptorOp, RLWENewDecryptorOp, RLWENewKeyGeneratorOp,
+                RLWEGenKeyPairOp, RLWEEncryptOp, RLWEDecryptOp,
+                BGVNewParametersFromLiteralOp, BGVNewEncoderOp,
+                BGVNewEvaluatorOp, BGVNewPlaintextOp, BGVEncodeOp, BGVDecodeOp,
+                BGVAddOp, BGVSubOp, BGVMulOp>(
+              [&](auto op) { return printOperation(op); })
+          .Default([&](Operation &) {
+            return emitError(op.getLoc(), "unable to find printer for op");
+          });
+
+  if (failed(status)) {
+    return emitError(op.getLoc(),
+                     llvm::formatv("Failed to translate op {0}", op.getName()));
+  }
+  return success();
+}
+
+LogicalResult LattigoEmitter::printOperation(ModuleOp moduleOp) {
+  os << kModulePreludeTemplate;
+
+  for (Operation &op : moduleOp) {
+    if (failed(translate(op))) {
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+LogicalResult LattigoEmitter::printOperation(func::FuncOp funcOp) {
+  // if (funcOp.getNumResults() != 1) {
+  //   return emitError(funcOp.getLoc(),
+  //                    llvm::formatv("Only functions with a single return type
+  //                    "
+  //                                  "are supported, but this function has ",
+  //                                  funcOp.getNumResults()));
+  //   return failure();
+  // }
+
+  os << "func " << funcOp.getName() << "() {\n";
+  os.indent();
+
+  for (Block &block : funcOp.getBlocks()) {
+    for (Operation &op : block.getOperations()) {
+      if (failed(translate(op))) {
+        return failure();
+      }
+    }
+  }
+
+  os.unindent();
+  os << "}\n";
+
+  return success();
+}
+
+LogicalResult LattigoEmitter::printOperation(func::ReturnOp op) {
+  // if (op.getNumOperands() != 1) {
+  //   return emitError(op.getLoc(), "Only one return value supported");
+  // }
+  // os << "return " << variableNames->getNameForValue(op.getOperands()[0])
+  //    << ";\n";
+  return success();
+}
+
+LogicalResult LattigoEmitter::printOperation(arith::ConstantOp op) {
+  auto valueAttr = op.getValue();
+  std::string valueString;
+  auto res = llvm::TypeSwitch<Attribute, LogicalResult>(valueAttr)
+                 .Case<IntegerAttr>([&](IntegerAttr intAttr) {
+                   valueString =
+                       "[]int64{" + std::to_string(intAttr.getInt()) + "}";
```

Comment:
I suspect `IntegerAttr` should generate a scalar value, while a `DenseAttr` or similar should generate an array value.

If we want to have a scalar `arith.constant` materialize as an array, it should be commented to explain why.

Reply:
```
func (ecd Encoder) Encode(values interface{}, pt *rlwe.Plaintext) (err error) {
```

In lattigo, `encoder.Encode` only accepts `[]int64` or `[]uint64` so I have to convert that to array.